### ISSUE TITLE
Not escaping ``&`` (ampersand) in entities breaks xml schema

### ISSUE THREAD

**Issue by makigumo** [thymeleaf/thymeleaf#255]
Considering the following source data, which should be displayed (e.g. in a textarea) as

``` xml
<a>b&lt;c&gt;</a>
```

which is escaped to

``` xml
&lt;a&gt;b&lt;c&gt;&lt;/a&gt;
```

and displayed via `th:text` as

``` xml
<a>b<c></a>
```

This is undesired and in this case simply wrong because it breaks the xml schema validation, because there is now opening markup for an element `c` defined which doesn't belong there. The correctly escaped sequence would be the following.

``` xml
&lt;a&gt;b&amp;lt;c&amp;gt;&lt;/a&gt;
```

Explicitly calling `#strings.escapeXml()` has the same problem, where I would expect it to behave the same as JSTLs `fn:escapeXml()` function.

Please provide a means to correctly escape the ampersand.

**Comment by lang** [comment]
It also affects thymeleaf-springs' th:field attribute: e.g. a "&amp;lt;" in a string property will be turned into "<" on rendering the form input.

**Comment by danielfernandez** [comment]
Actually, after much research, it seemed that this escaping bug was not only quite deeply rooted into thymeleaf's way of dealing with `Text` and `Attribute` nodes, but also very difficult to solve with the existing java escaping tools.

The reason for this is that Thymeleaf needs to:
1. Unescape the contents of the template texts and attributes that will be processed.
2. Process them, adding data from databases and such if needed.
3. Escape texts and attributes again for output.

**But** given the existing _unescape_ operation from step 1 wasn't really complete because it only unescaped the basic --or _markup-significant_-- set of escapes (`&lt;`, `&gt;`, `&amp;`, `&quot;` and `&apos;`), whenever the _escape_ operation in step 3 encountered something like `&euro;`, it couldn't know whether it came from:
1. The string `&amp;euro;` in template input.
2. The string `&euro;` in template input.
3. The string `&euro;` from database or similar.

So, this escape operation in step 3 was only doing the inverse to step 1, i.e. only escaping `<`, `>`, `&`, `"` and `'`. And therefore _ignoring on purpose any already-escaped HTML character references_ like `&euro;`.. or your `&lt;` and `&gt;`.

That's why it was outputting:

```
&lt;a&gt;b&lt;c&gt;&lt;/a&gt;
```

...instead of the correct:

```
&lt;a&gt;b&amp;lt;c&amp;gt;&lt;/a&gt;
```

**How to solve this?**

Not really easy, because the only way to make sure that the result of the thymeleaf "escape" operation in step 3 is correct is to also ensure that the previous "unescape" in step 1 has been correct and complete, and therefore no character references like `&euro;`, `&lt;`, `&nbsp;` or even `&#xE1;` have been left escaped from input. That way, whenever `&euro;` is found in step 3, there would be no doubt that it has to be escaped as `&amp;euro;`, because if not it would have been found as `€`.

So thymeleaf needed a more powerful HTML unescape operation. **Much** more powerful. Because HTML4 defines 252 character references, but it happens that HTML5 defines more than 2,000. And none of the existing tools did such a complete unescape operation (Commons-Lang, Spring's `HtmlUtils`, etc..)

So... I had to develop it myself. I created **Unbescape** [ http://www.unbescape.org ] in order to being able to perform complete, correct and high-performance HTML, XML, JavaScript, Java, CSS and CSV operations, with the aim of making thymeleaf use it and avoid escaping problems once and for all.

**Conclusion**

The thymeleaf core has been modified in order to use Unbescape for all escape/unescape operations (including the JavaScript ones when doing JavaScript inlining). This will be a part of 2.1.3.

2.1.3-SNAPSHOT has been updated with this feature, so this can be tested now by using this snapshot version.
